tools/pepsi/src/player_number.rs

```diff
@@ -45,7 +45,15 @@ pub async fn player_number(arguments: Arguments, repository: &Repository) -> Res
         arguments.assignments,
         "Setting player number...",
         |assignment, _progress_bar| {
-            let head_id = &hardware_ids[&assignment.nao_number.number].head_id;
+            let hardware = &hardware_ids
+                .get(&assignment.nao_number.number)
+                .unwrap_or_else(|| {
+                    panic!(
+                        "NAO with Hardware ID {} does not exist",
+                        &assignment.nao_number.number
+                    )
```

Comment:
Please return a proper error instead of panic here

Reply:
What's the advantage?

Reply:
Same question from me. But additionally if I did want to return a proper error here, how would I do this since the Result<_> does not implement IntoFuture and I therefore cannot return the Error out pf map_tasks back into the player_number function?

Reply:
This indeed needs some adjustment to the `map_tasks` function, as this falsely required the closure to be `Copy`. See below...

```
pub async fn map_tasks<T, F, M>(
    items: impl IntoIterator<Item = T>,
    message: &'static str,
    task: impl Fn(T, ProgressBar) -> F,
) where
    T: ToString,
    F: Future<Output = Result<M>>,
    M: Into<TaskMessage>,
{
    let multi_progress = Self::new();
    items
        .into_iter()
        .map(|item| (multi_progress.task(item.to_string()), item))
        .map(|(progress, item)| {
            progress.enable_steady_tick();
            progress.set_message(message);
            let future = task(item, progress.progress.clone());
            async move { progress.finish_with(future.await) }
        })
        .collect::<FuturesUnordered<_>>()
        .collect::<Vec<_>>()
        .await;
}
```

```
let hardware_ids = &hardware_ids;
ProgressIndicator::map_tasks(
    arguments.assignments,
    "Setting player number...",
    |assignment, _progress_bar| async move {
        let number = assignment.nao_number.number;
        let hardware_id = hardware_ids
            .get(&number)
            .ok_or_else(|| eyre!("NAO with Hardware ID {number} does not exist",))?;
        repository
            .set_player_number(&hardware_id.head_id, assignment.player_number)
            .await
            .wrap_err_with(|| format!("failed to set player number for {assignment}"))
    },
)
.await;
```

Reply:
Thank you! This indeed much prettier. I have implemented the changes as recommended

Reply:
Done